### PR TITLE
Introduce .toNumber()

### PR DESCRIPTION
**Pull request by axic**
A simple wrapper around toString() to get the an actual number object in JS. Throws an error if its not safe to represent in JS.

### CODE REVIEW

File: lib/bn.js [hunk 1]
```diff
@@ -450,6 +450,11 @@ BN.prototype.toString = function toString(base, padding) {
   }
 };
 
+BN.prototype.toNumber = function toNumber() {
+  assert(this.bitLength() <= 53, 'Number can only safely store up to 53 bits');
+  return parseInt(this.toString());
```
Comment (indutny) [inline lib/bn.js]: `parseInt(..., 10)`?

**Comment by indutny** [comment]
And use case is?

File: lib/bn.js (hunk 1 above)
Comment (axic) [inline lib/bn.js]: That will allow bitLength()>53 numbers to be wrongly transformed into number. JS won't complain, see http://stackoverflow.com/questions/307179/what-is-javascripts-highest-integer-value-that-a-number-can-go-to-without-losin

**Comment by axic** [comment]
It is a shorthand to parseInt + BN.toString with overflow checks.

I'm using it in a library for handling Ethereum's ABI.

File: lib/bn.js (hunk 1 above)
Comment (indutny) [inline lib/bn.js]: I wasn't suggesting to remove `assert`

**Comment by indutny** [comment]
I'm not sure if it is really useful for everyone... cc @dcousens

**Comment by indutny** [comment]
cc @calvinmetcalf

File: lib/bn.js (hunk 1 above)
Comment (axic) [inline lib/bn.js]: Ah, I've assumed parseInt is radix 10 by default, but it is radix 8 if it starts with '0'?

Comment (indutny) [inline lib/bn.js]: Exactly.

Comment (axic) [inline lib/bn.js]: toString() defaults to radix 10 and padding 1. In what case would it result in leading zeroes (apart from the number 0)?

Comment (indutny) [inline lib/bn.js]: Nope, it won't result in leading `0`

**Comment by calvinmetcalf** [comment]
I think this could be useful, for extracting a number and avoiding accidentally getting a result which is not as precise as you think.

**Comment by indutny** [comment]
This patch does not apply cleanly to master. May I ask you to run:

```
git fetch origin
git rebase origin/master
```

And resolve all conflicts?

**Comment by dcousens** [comment]
Couldn't you just use `parseInt`?
Could that be less accurate than expected?

**Comment by axic** [comment]
@dcousens it was changed to parseInt(.., 10) on the request of @indutny assuming that in some hypothetical cases toString() could return with a leading zero and an ancient Javascript interpreter might consider that as octal. Probably safer this way.

**Comment by axic** [comment]
@indutny did the rebase

**Comment by indutny** [comment]
It was still messed up, but I figured it out ;)

Thanks, landed!

**Comment by axic** [comment]
I need to clean up my tree as the previous pull request was cherry picked and that seems to have messed it up.

**Comment by indutny** [comment]
Yeah, I always cherry-pick.

**Comment by axic** [comment]
Thank you for the merge! I'll just recreate my fork to avoid any future mess as there's no other pull request pending.

**Pull request closed without merging**